samcli/lib/cookiecutter/template.py

```diff
@@ -0,0 +1,158 @@
+            the values of the cookiecutter.json keys, i.e. the cookiecutter context
+        preprocessors: Optional[List[Processor]]
+            An optional series of processes to be executed before generating the cookiecutter project. These processes
+            are usually used to manipulate the cookiecutter context or doing some bootstrapping before generating
+            the cookiecutter project.
+        postprocessors: Optional[List[Processor]]
+            An optional series of processes to be executed after generating the cookiecutter project. These processes
+            are usually used to clean up resources and to print messages to the user after successfully generating
+            the cookiecutter project.
+        plugins: Optional[List[Plugin]]
+            An optional series of plugins to be plugged in. A plugin defines its own interactive_flow, preprocessor
+            and postprocessor. A plugin is a sub-set of the template, if there is a common behavior among multiple
+            templates, it is better to be extracted to a plugin that can then be plugged in to each of these templates.
+        """
+        self._location = location
+        self._interactive_flows = interactive_flows or []
+        self._preprocessors = preprocessors or []
+        self._postprocessors = postprocessors or []
+        self._plugins = plugins or []
+        for plugin in self._plugins:
+            if plugin.interactive_flow:
+                self._interactive_flows.append(plugin.interactive_flow)
+            if plugin.preprocessor:
+                self._preprocessors.append(plugin.preprocessor)
+            if plugin.postprocessor:
+                self._postprocessors.append(plugin.postprocessor)
+
+    def run_interactive_flows(self) -> Dict:
+        """
+        prompt the user a series of questions' flows and gather the answers to create the cookiecutter context.
+        The questions are identified by keys. If multiple questions, whether within the same flow or across
+        different flows, have the same key, the last question will override the others and we will get only the
+        answer of this question.
+
+        Raises:
+            UserException(ClickException) if anything went wrong.
+
+        Returns:
+            A Dictionary in the form of {question.key: answer} representing user's answers to the flows' questions
+        """
+        try:
+            context: Dict[str, Any] = {}
+            for flow in self._interactive_flows:
+                context = flow.run(context)
+            return context
+        except Exception as e:
+            raise UserException(str(e), wrapped_from=e.__class__.__name__) from e
+
+    def generate_project(self, context: Dict):
+        """
+        Generates a project based on this cookiecutter template and the given context. The context is first
+        processed and manipulated by series of preprocessors(if any) then the project is generated and finally
```

Comment:
Q: Where are the pre-processors and post-processors defined, and what kind of actions are done? are they controlled by us or by users?

Reply:
This is the workflow. Any template/plugin uses this workflow will define its own preprocessing and postprocessing logic, if needed. The next PR will have an example plugin using it.